### PR TITLE
add tracer particles

### LINKED ISSUE #489: add tracer particles
**Describe the proposal**
Add tracer particles that advect with the flow, using the face-centered velocities returned by the Riemann solver, following the `Advection_AmrCore` example code (https://github.com/AMReX-Codes/amrex/blob/e0b77e1fce491021696d22f1496138e8d9942d65/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp).

**Additional context**
Tracer particles are easy to add, since the AMReX `AmrTracerParticleContainer` can be used: https://github.com/AMReX-Codes/amrex/blob/e0b77e1fce491021696d22f1496138e8d9942d65/Src/AmrCore/AMReX_AmrParticles.H#L293

Tracer particles are never deposited onto a grid, so it is unnecessary to have ghost/virtual particles. This makes the implementation easy. The only non-trivial steps are to call `TracerPC->Redistribute` after particles are advected and when regridding happens.

The Castro implementation (using cell-centered velocities) is [here](https://github.com/AMReX-Astro/Castro/blob/main/Source/particles/CastroParticles.cpp).

The original face-centered velocity tracer particle implementation is [Harlow & Welch (1965)](https://ui.adsabs.harvard.edu/abs/1965PhFl....8.2182H/abstract).

### PR DESCRIPTION
**Pull request by BenWibking**
### Description
This adds tracer particles. Particles are created one-per-cell in the initial conditions, then they advect with the fluid using the face-centered velocities returned from the Riemann solver. Particles are included in plotfiles and checkpoints. *(Particles will be added to openPMD outputs in a subsequent PR.)*

The method adopted here for particle advection with face-centered velocities is the ["marker-and-cell method" (Harlow & Welch 1965)](https://ui.adsabs.harvard.edu/abs/1965PhFl....8.2182H/abstract).

Tracer particles are turned on for the `HydroBlast3D` and `SphericalCollapse` problems, so tracer particle functionality is included in the test suite.

*Significant code infrastructure changes:*
* This removes the old AMR level subcycling code taken from Chombo and replaces it with the cleaner version from the AMReX code examples. However, it should behave identically as before. This is necessary for particle redistribution to work with subcycling.
* This adds the face-centered velocities to `state_new_fc_` and `state_old_fc_`. Some changes to the machinery for face-centered variables was necessary to get this to work, such as implementing AMR interpolation for face-centered variables using the `amrex::FaceLinear` interpolator (*caution: this interpolator does NOT preserve the divergence of the field!*). This is necessary because ghost faces are needed in order to handle particles that cross coarse-fine boundaries during a timestep. (In the future, this machinery can be modified so that it creates alias MultiFabs for separate face-centered components and then uses a different AMR interpolator for each component.)
* In order to avoid modifying the initial conditions for every problem when tracer particles are enabled, this separates the initial conditions set-up function into `setInitialConditionsOnGrid` (cell-centered variables) and `setInitialConditionsOnGridFaceVars` (face-centered variables). This only affects the `FCQuantities` test problem.

### Related issues
Closes https://github.com/quokka-astro/quokka/issues/489 and https://github.com/quokka-astro/quokka/issues/498.

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [x] I have added a description (see above).
- [x] I have added a link to any related issues see (see above).
- [x] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [x] I have added tests for any new physics that this PR adds to the code.
- [x] I have tested this PR on my local computer and all tests pass.
- [x] I have manually triggered the GPU tests with the magic comment `/azp run`.
- [x] I have requested a reviewer for this PR.

### CODE REVIEW

File: src/RadhydroSimulation.hpp [hunk 1]
```diff
@@ -661,6 +663,38 @@ template <typename problem_t> void RadhydroSimulation<problem_t>::advanceSingleT
...
+			amrex::Real vy = state[bx](i, j, k, HydroSystem<problem_t>::x2Momentum_index) / rho;
+			amrex::Real vz = state[bx](i, j, k, HydroSystem<problem_t>::x3Momentum_index) / rho;
+			vel[bx](i, j, k, 0) = vx;
+			vel[bx](i, j, k, 1) = vy;
+			vel[bx](i, j, k, 2) = vz;
+		});
+		amrex::Gpu::streamSynchronizeAll();
+
+		// fill ghost velocity cells
+		vel_mf.FillBoundary(geom[lev].periodicity());
```
Comment (BenWibking) [inline src/RadhydroSimulation.hpp]: This might give inconsistent velocities at the coarse-fine boundaries.

**Comment by BenWibking** [comment]
There is a weird ROCm build issue:

```
In file included from /home/runner/work/quokka/quokka/src/RadMatterCouplingRSLA/test_radiation_matter_coupling_rsla.cpp:10:
In file included from /home/runner/work/quokka/quokka/src/RadMatterCouplingRSLA/test_radiation_matter_coupling_rsla.hpp:21:
In file included from /home/runner/work/quokka/quokka/src/radiation_system.hpp:29:
In file included from /home/runner/work/quokka/quokka/src/hyperbolic_system.hpp:35:
In file included from /home/runner/work/quokka/quokka/src/simulation.hpp:76:
In file included from /home/runner/work/quokka/quokka/extern/amrex/Src/AmrCore/AMReX_AmrParticles.H:5:
In file included from /home/runner/work/quokka/quokka/extern/amrex/Src/Particle/AMReX_Particles.H:5:
In file included from /home/runner/work/quokka/quokka/extern/amrex/Src/Particle/AMReX_ParticleContainer.H:1406:
In file included from /home/runner/work/quokka/quokka/extern/amrex/Src/Particle/AMReX_ParticleIO.H:5:
/home/runner/work/quokka/quokka/extern/amrex/Src/Particle/AMReX_WriteBinaryParticleData.H:161:14: error: reference to __host__ function 'memcpy' in __host__ __device__ function
        std::memcpy(&yi, &yu, sizeof(yu));
             ^
2 errors generated when compiling for gfx908.
```

**Comment by BenWibking** [comment]
unrelated macOS CI failure.

**Comment by BenWibking** [comment]
On the KelvinHelmholz problem with AMR, it crashes about halfway through with the error:
```
[Bens-MacBook-Pro:93170] [ 0] libc++abi: terminating due to uncaught exception of type std::runtime_error: ParticleContainer::locateParticle(): invalid particle.
```

**Comment by BenWibking** [comment]
~~reliably crashes when AMR is enabled. need to debug~~

Fixed. Needs at least 2 ghost cells for the face velocity.

**Comment by BenWibking** [comment]
On uniform grids, it seems to work fine:

kh_tracer

**Comment by BenWibking** [comment]
This AMReX PR should fix the ROCm build issue: https://github.com/AMReX-Codes/amrex/pull/3699

**Comment by BenWibking** [comment]
@markkrumholz There are a lot of code infrastructure changes in this PR, unfortunately. Let me know whether I should split this up into multiple PRs, and if so, what splitting would be easiest to review.

File: src/AdvectionSimulation.hpp
Comment (markkrumholz) [inline src/AdvectionSimulation.hpp:1]: Would probably be good to modify the comment here to note that an implementation is only required if tracer particles are enabled.

File: src/HydroBlast2D/test_hydro2d_blast.cpp
Comment (markkrumholz) [inline src/HydroBlast2D/test_hydro2d_blast.cpp:1]: What is the purpose of this change? It seem unrelated to tracer particles. Should it go into a separate PR?

File: src/HydroKelvinHelmholz/test_hydro2d_kh.cpp
Comment (markkrumholz) [inline src/HydroKelvinHelmholz/test_hydro2d_kh.cpp:1]: Again, these changes seem unrelated to tracer particles. Should they be in this PR?

File: src/HydroBlast2D/test_hydro2d_blast.cpp
Comment (BenWibking) [reply src/HydroBlast2D/test_hydro2d_blast.cpp:1]: Oops, that was for debugging. I'll remove those.

File: src/HydroKelvinHelmholz/test_hydro2d_kh.cpp
Comment (BenWibking) [reply src/HydroKelvinHelmholz/test_hydro2d_kh.cpp:1]: I'll remove this.

File: src/RadhydroSimulation.hpp [hunk 2]
```diff
@@ -857,6 +869,14 @@ void RadhydroSimulation<problem_t>::advanceHydroAtLevelWithRetries(int lev, amre
 		amrex::Copy(originalFineData, fineData, 0, 0, fineData.nComp(), 0);
 	}
 
+#ifdef AMREX_PARTICLES
```
Comment (markkrumholz) [inline src/RadhydroSimulation.hpp:873]: I don't understand the logic for using the AMREX_PARTICLES guards here. Should this be controlled by physics traits like everything else? Why use ifdef's here instead?

File: src/simulation.hpp [hunk 3]
```diff
@@ -72,6 +73,11 @@ namespace filesystem = experimental::filesystem;
 #include <fmt/core.h>
 #include <yaml-cpp/yaml.h>
 
+#ifdef AMREX_PARTICLES
```
Comment (markkrumholz) [inline src/simulation.hpp:76]: Same comment as above about the AMREX_PARTICLES guards.

File: src/simulation.hpp [hunk 4]
```diff
@@ -316,7 +355,7 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
 
 	// Nghost = number of ghost cells for each array
 	int nghost_cc_ = 4; // PPM needs nghost >= 3, PPM+flattening needs nghost >= 4
-	int nghost_fc_ = 4;
```
Comment (markkrumholz) [inline src/simulation.hpp:319]: How does this change interact with MHD? Presumably MHD needs 4, so this number should depend on physics_traits.

**Review by markkrumholz — COMMENTED** [review]
I've left a few comments on individual files, but overall this looks fine. Once those comments are addressed I can approve.

File: src/RadhydroSimulation.hpp (hunk 2 above)
Comment (BenWibking) [reply src/RadhydroSimulation.hpp:873]: It's possible to compile AMReX without particle support. That might conceivably be useful for debugging (e.g., compiler bugs) in the future, so I'd like to keep it with the ifdef here. Otherwise, it is entirely controlled with the runtime parameter `do_tracers`.

File: src/simulation.hpp (hunk 4 above)
Comment (BenWibking) [reply src/simulation.hpp:319]: I'll change this so it uses 4 if MHD is enabled.

**Comment by BenWibking** [comment]
@markkrumholz Can you re-review?

File: src/RadhydroSimulation.hpp (hunk 2 above)
Comment (markkrumholz) [reply src/RadhydroSimulation.hpp:873]: OK, that's fine.

**Review by markkrumholz — APPROVED** [review]
OK, looks good now.
